### DISCUSSION TITLE
RTU READ_HOLD_REGISTER Once Worked Now Doesn't

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [eModbus/eModbus#279] [Q&A]
I'm not sure if this is related to version 1.7.0 release but it may be (I haven't had a chance to downgrade the library and check).

I was running basically the example code for RTU04 with some minor changes.  I noticed recently that I wasn't getting any MODBUS data and I saw mostly `E0 - Timeout` and `E5 - Packet Length error` messages. I'm only reading one register and 1 byte.

If I run QModMaster on my PC and connect to the board I'm trying to read I can see the correct traffic:

```
[RTU]>Tx > 15:04:02:461 - 64  03  00  00  00  01  8D  FF  
[RTU]>Rx > 15:04:02:473 - 64  03  02  00  00  F4  4C  
[RTU]>Tx > 15:04:12:609 - 64  03  00  00  00  01  8D  FF  
[RTU]>Rx > 15:04:12:621 - 64  03  02  00  07  B5  8E  
```
Here the register was holding `0` the first time and then `7` the 2nd time.

When I connect my ESP32 board running my code and enable debugging I see this:

```
[D] 68511| ModbusClientRTU.cpp  [ 147] addRequestM: request for 64/03
[D] 68511| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01
[D] 68511| ModbusClientRTU.cpp  [ 158] addRequestM: RC=00
[D] 68512| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB4678/6:
  | 0000: 64 03 00 00 00 01                                 |d.....          |
[D] 68536| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[D] 70548| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FFB426C/1:
  | 0000: E0                                                |.               |
[D] 70558| ModbusClientRTU.cpp  [ 268] handleConnection: Error response (1 bytes) received.
[D] 70569| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[LOGFS][WARNING]::[src/modbus.cpp:38]:Error response: E0 - Timeout


[D] 38496| ModbusClientRTU.cpp  [ 147] addRequestM: request for 64/03
[D] 38497| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01
[D] 38497| ModbusClientRTU.cpp  [ 158] addRequestM: RC=00
[D] 38497| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB4678/6:
  | 0000: 64 03 00 00 00 01                                 |d.....          |
[D] 38522| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[D] 38535| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FFB426C/1:
  | 0000: E5                                                |.               |
[D] 38545| ModbusClientRTU.cpp  [ 268] handleConnection: Error response (1 bytes) received.
[D] 38556| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[LOGFS][WARNING]::[src/modbus.cpp:38]:Error response: E5 - Packet length error
```

I saw that you had to add a setup before beginning your Serial port so I added the `RTUutils`:
```
 // Set up Serial2 connected to Modbus RTU
    RTUutils::prepareHardwareSerial(Serial2);
    Serial2.begin(BAUDRATE, SERIAL_8N1, RXPIN, TXPIN);
```

I didn't capture the debugging before adding this line but the same errors are happening after the change as they were before.

Perhaps I'm missing something "obvious" that I need to add to my code but I know at some point the code and boards were working so I thought I'd post here while I continue to try and debug.

**Comment by Miq1** [comment]
That is odd indeed. The changes in 1.7.0 were intended to exactly avoid the error pattern you are experiencing.

The received packets are of length 1 because they do not have any data except the error code byte, so this at least is explainable.

To see what was received in the background you should set a ``LOG_LEVEL`` of 6 (or ``LOG_LEVEL_VERBOSE`` when building. Then the raw data will be logged as well.

If you will post the verbose lig here, we may see what is happening.

**Comment by GeorgeIoak** [comment]
OK, I'll recompile with `LOG_LEVEL_VERBOSE` and post the log. I hooked up the logic analyzer to the A/B signals and I could see the same request sent and there was a response but it didn't start with `64 03`. I thought about hooking up the LA on the UART side of the transceiver of both boards but I'll do the `LOG_LEVEL_VERBOSE` first and see if what it's receiving looks the same as what the LA decoded.

**Reply by Miq1** [reply]
:+1 - looking forward to it!

**Comment by GeorgeIoak** [comment]
OK, here's a couple of log captures:

```
V] 43348| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 43348| ModbusClientRTU.cpp  [ 147] addRequestM: request for 64/03
[D] 43359| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01
[D] 43359| ModbusClientRTU.cpp  [ 158] addRequestM: RC=00
[D] 43359| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB5DD4/6:
  | 0000: 64 03 00 00 00 01                                 |d.....          |
[D] 43381| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[V] 43393| RTUutils.cpp         [ 261] receive: C/1827us without data after 1
[V] 43393| RTUutils.cpp         [ 271] receive: C/[V] Raw buffer received: @3FFAF008/1:
  | 0000: 00                                                |.               |
[D] 43414| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FFB5E34/1:
  | 0000: E5                                                |.               |
[D] 43424| ModbusClientRTU.cpp  [ 268] handleConnection: Error response (1 bytes) received.
[V] Data: @3FFB5E34/1:
  | 0000: E5                                                |.               |
[D] 43445| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[V] Response packet: @3FFB5E34/3:
  | 0000: 64 83 E5                                          |d..             |
[LOGFS][WARNING]::[src/modbus.cpp:38]:Error response: E5 - Packet length error
```

and I also see:

```
[V] 403544| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 403544| ModbusClientRTU.cpp  [ 147] addRequestM: request for 64/03
[D] 403555| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01
[D] 403555| ModbusClientRTU.cpp  [ 158] addRequestM: RC=00
[D] 403555| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB5DD4/6:
  | 0000: 64 03 00 00 00 01                                 |d.....          |
[D] 403576| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[V] 403588| RTUutils.cpp         [ 261] receive: C/1827us without data after 6
[V] 403588| RTUutils.cpp         [ 271] receive: C/[V] Raw buffer received: @3FFAF034/6:
  | 0000: 0F 02 00 F8 7C FF                                 |....|.          |
[D] 403609| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FFB5E34/1:
  | 0000: E2                                                |.               |
[D] 403619| ModbusClientRTU.cpp  [ 268] handleConnection: Error response (1 bytes) received.
[V] Data: @3FFB5E34/1:
  | 0000: E2                                                |.               |
[D] 403640| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[V] Response packet: @3FFB5E34/3:
  | 0000: 64 83 E2                                          |d..             |
[LOGFS][WARNING]::[src/modbus.cpp:38]:Error response: E2 - CRC check error
```

I've got the LA connected up this morning to 3 spots, UART side of RS-485 transceiver of both boards as well as the A/B signals. This MAY be influencing the results.

- I know that the board I'm trying to read works when the USB-to-RS-485 adapter is connected to the PC and I run software on the PC so that should eliminate any hardware problem on the board I'm trying to read.
- I know that using the LA and probing the A/B line I at least saw the correct `64  03  00  00  00  01  8D  FF` being sent but I haven't seen the correct response in my LA captures. I haven't gone back and connected the LA when the board is connected to the PC rather than my ESP32 running eModBus but I assume the data should show up since the PC receives the correct response.

**Reply by Miq1** [reply]
The first log looks like a 'ghost' 00 byte that can be generated by the RE/DE switching logic. I assume you are using RS485 adapters with a RE/DE line that needs to be toggled?

The second is completely unclear to me. The data captured have no resemblance to those expected.

A few basics: how long is your wiring? Do you have proper termination resistors on both and biasing resistors on one end of the RS485 line?

If you link your LA to the UART RX and TX pins, do you see the correct responses? That is not clear to me from your descriptiin.

**Reply by GeorgeIoak** [reply]
The RS-485 transceivers ([HG Semi MAX3072E](https://datasheet.lcsc.com/lcsc/2005111332_HGSEMI-MAX3072EIM-TR_C512782.pdf)) are on on both boards and yes they use RE/DE. The circuit is this:
![image](https://user-images.githubusercontent.com/1467145/225408029-95520860-49d0-409f-81db-a4f8ded3594d.png)
 with 485_EN going to ESP32 GPIO15, UART1_Tx to GPIO32, and UART1_Rx to GPIO34

The wiring is maybe 6 inches long. When I connect the board I'm trying to read to the computer and successfully read from it the wires are also ~6 inches long. I'm not sure what the exact transceiver chip used for the USB-to-RS-485 Adapter.

With my ESP32 connected to the board we're reading I have not seen a correct response on the LA coming from the board. I have not connected the LA to the board when it is read from the PC but I could. I just figured if the PC confirms the correct response then the LA would show the same although if it's a slight timing problem perhaps the PC software is more forgiving than eModBus. 

From the `VERBOSE` logs it looks like at some point we see `64 83 E2` so the address matches correctly but nothing after that.

**Reply by bertmelis** [reply]
What is the voltage between A and B when the bus is idle?

**Reply by GeorgeIoak** [reply]
-0.9VDC with GND on the A terminal of the board I'm trying to read and the ESP32 board disconnected

**Reply by GeorgeIoak** [reply]
When I connect "the board" to my PC by itself (ESP32 board is disconnected) and read with the PC (via the USB-to-RS-485 Adapter) the LA correctly sees the response when I am reading the UART side of the chip on the board (the board has a STM32 as the main processor). 

When I check the A/B terminals I can see the correct request message (`64 03 00 00 00 01 0D 7F` not sure why CRC is 0D7F and before it was 8DFF) but the response does not match what the PC displays:

- PC sends `64 03 00 00 00 01 0D 7F 7F` according to the LA but the PC software says it sent `64 03 00 00 00 01 8D FF`
- PC receives `1B 79 7B 7F 7F 17 56` according to the LA but the PC software says it received `64 03 02 00 00 F4 4C`

When I switch the LA back to the UART side:

- PC sends `64 03 00 00 00 01 0D 7F` according to the LA and the PC software says it sent `64 03 00 00 00 01 8D FF`
- PC receives `64 03 02 00 00 74 4C` according to the LA and the PC software says it received `64 03 02 00 00 F4 4C`

I've never sniffed the RS-485 signals before but maybe the levels are a little low for the LA to read reliably? One thing for sure, this is rather curious.

**Reply by Miq1** [reply]
> When I check the A/B terminals I can see the correct request message (64 03 00 00 00 01 0D 7F not sure why CRC is 0D7F and before it was 8DFF) but the response does not match what the PC displays:

The CRC you see seems to be displayed as 7 bit ASCII only, the top bit is missing!

There is something wrong with your LA's settings I reckon.

> From the VERBOSE logs it looks like at some point we see 64 83 E2 so the address matches correctly but nothing after that.

The response is generated locally in the eModbus lib from the error detected, so no wonder the server ID etc. is correct 😉

**Reply by GeorgeIoak** [reply]
Yeah, sorry, I just caught that the LA setting was 7-bit instead of 8-bit so that's fixed now.

**Comment by GeorgeIoak** [comment]
OK, not good news for me but the plot thickens.

I took my ESP32 board and connected it to my PC using that same USB-to-RS485 Adapter and this time I ran pyModSlave software. Looks like the ESP32 eModBus correctly read the PC:

```
[V] 33841| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 33841| ModbusClientRTU.cpp  [ 147] addRequestM: request for 64/03
[D] 33852| ModbusClientRTU.cpp  [ 230] addToQueue: RC=01
[D] 33852| ModbusClientRTU.cpp  [ 158] addRequestM: RC=00
[D] 33852| ModbusClientRTU.cpp  [ 248] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB633C/6:
  | 0000: 64 03 00 00 00 01                                 |d.....          |
[D] 33878| ModbusClientRTU.cpp  [ 253] handleConnection: Request sent.
[V] 33908| RTUutils.cpp         [ 261] receive: C/1831us without data after 7
[V] 33908| RTUutils.cpp         [ 271] receive: C/[V] Raw buffer received: @3FFAEFA0/7:
  | 0000: 64 03 02 00 05 34 4F                              |d....4O         |
[D] 33918| RTUutils.cpp         [ 436] receive: C/[D] Received packet: @3FFB6350/5:
  | 0000: 64 03 02 00 05                                    |d....           |
[D] 33939| ModbusClientRTU.cpp  [ 268] handleConnection: Data response (5 bytes) received.
[V] Data: @3FFB6350/5:
  | 0000: 64 03 02 00 05                                    |d....           |
[D] 33950| ModbusClientRTU.cpp  [ 289] handleConnection: Response generated.
[V] Response packet: @3FFB6350/5:
  | 0000: 64 03 02 00 05                                    |d....           |
```

So it seems that each board works as expected when connected to a PC running MODBUS software but the 2 boards can't talk to each other. What could possibly cause something like this?

**Reply by Miq1** [reply]
Line levels, most probably. 

The PC may pull it to a reasonable level, even if the "board" is trying to numb it, but the ESP32 is too weak to achieve that. Theory only, you are at the spot. 

Try using different or exchanging the RS485 adapters to see if the error pattern will move with an adapter erc.

**Reply by GeorgeIoak** [reply]
Yeah, looks like I've got more probing to do to figure this out

**Reply by bertmelis** [reply]
The 10k pullup/down resistors might be too high. Try lowering it to let's say 1k.

**Reply by GeorgeIoak** [reply]
You're talking about on the A/B lines, right?

**Reply by bertmelis** [reply]
yes

**Reply by GeorgeIoak** [reply]
Just as a sanity check I thought I'd check what the signal looked like on the scope. Here's the ESP32 A terminal with the scope ground on the B terminal. The levels coming from the PC were ~3.3V as opposed to the ESP32 ~2.5V. Not sure  if this is what they typically look like or if the level or slow rise to peak looks abnormal. Changing the PU/PD from 10k to 1k would help these edges but I'm thinking this doesn't look too bad?
![image](https://user-images.githubusercontent.com/1467145/225439093-1d36f782-f366-4ae0-ba31-f88ea8b59de2.png)

**Reply by Miq1** [reply]
For me the signal does not look very good, to be honest. I would expect the flanks to be much more straight. It almost seems like there is a capacitor hidden.

**Reply by bertmelis** [reply]
I don't have a scope, so I can't relate. But it doesn't look very odd to me. Don't know what the LA makes from this.

You'll have to check your driver's datasheet to see if the levels are OK.

But lowering the PU/PD will create a stronger PU/PD.
